src/Proxy.NET/Authentication/ApiKeyAuthenticationHandler.cs

```diff
@@ -0,0 +1,32 @@
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.Extensions.Options;
+using Proxy.NET.Services;
+
+namespace Proxy.NET.Authentication;
+
+public class ApiKeyAuthenticationHandler(
+    IOptionsMonitor<ProxyAuthenticationOptions> options,
+    IAuthorizeService authorizeService,
+    ILoggerFactory logger,
+    UrlEncoder encoder
+) : AuthenticationHandler<ProxyAuthenticationOptions>(options, logger, encoder)
+{
+    protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+    {
+        if (!Request.Headers.TryGetValue("api-key", out var apiKey) || string.IsNullOrEmpty(apiKey))
+            return AuthenticateResult.Fail("Authentication failed.");
+
+        var requestContext = await authorizeService.IsUserAuthorized(apiKey!);
```

Comment:
```suggestion
        var requestContext = await authorizeService.IsUserAuthorized(apiKey);
```

The compiler should pick up that `apiKey` isn't `null` due to the check above.